crates/polars-stream/src/nodes/reduce.rs

```diff
@@ -0,0 +1,205 @@
+                }
+            }
+
+            full.lock().push(reductions);
+
+            Ok(())
+        })
+    }
+
+    fn spawn_source<'env, 's>(
+        &'env self,
+        scope: &'s TaskScope<'s, 'env>,
+        _pipeline: usize,
+        _recv: &mut [Option<Receiver<Morsel>>],
+        send: &mut [Option<Sender<Morsel>>],
+        _state: &'s ExecutionState,
+    ) -> JoinHandle<PolarsResult<()>> {
+        let ReduceState::Source(df) = &self.state else {
+            unreachable!()
+        };
+        let mut send = send[0].take().unwrap();
+
+        scope.spawn_task(TaskPriority::High, async move {
+            let Some(df) = df.lock().take() else {
+                return Ok(());
+            };
+            let morsel = Morsel::new(df, MorselSeq::new(0));
+            let _ = send.send(morsel).await;
+            Ok(())
+        })
+    }
+}
+
+impl ComputeNode for ReduceNode {
+    fn name(&self) -> &str {
+        "reduce"
+    }
+
+    fn update_state(&mut self, recv: &mut [PortState], send: &mut [PortState]) {
+        assert!(recv.len() == 1 && send.len() == 1);
+
+        // State transitions
+        // If the output doesn't want any more data, transition to being done.
+        if send[0] == PortState::Done && !matches!(&self.state, ReduceState::Done) {
+            self.state = ReduceState::Done;
+        }
+
+        match self.state {
+            // Input is done, we can combine the reductions into a single scalar.
+            ReduceState::Sink { .. } if matches!(recv[0], PortState::Done) => {
+                let reductions = std::mem::take(&mut *self.full.lock());
+
+                let reductions = reductions
+                    .into_iter()
+                    .map(PolarsResult::Ok)
+                    .reduce(|a, b| {
+                        let mut a = a?;
+                        let mut b = b?;
+                        for (a, b) in a.iter_mut().zip(b.iter_mut()) {
+                            a.combine(b.as_ref())?
+                        }
+                        Ok(a)
+                    })
+                    .expect("at least on thread");
+
+                // TODO! make `update_state` fallible.
+                let reductions = reductions.unwrap();
+
+                let columns = reductions
+                    .into_iter()
+                    .zip(self.output_schema.iter_fields())
+                    .map(|(mut r, field)| {
+                        r.finalize().map(|scalar| {
+                            scalar.into_series(&field.name).cast(&field.dtype).unwrap()
+                        })
+                    })
+                    .collect::<PolarsResult<Vec<_>>>()
+                    .unwrap();
+                let out = unsafe { DataFrame::new_no_checks(columns) };
+
+                self.state = ReduceState::Source(Mutex::new(Some(out)));
+            },
+            // We have fed the source, we are done.
+            ReduceState::Source(..) => {
```

Comment:
This state transition is not correct, you need to check whether the `Option` inside the `Mutex` is actually `None`, that is, we actually sent the value.